test/core/end2end/h2_ssl_session_reuse_test.cc

```diff
@@ -263,9 +243,93 @@ TEST(H2SessionReuseTest, SingleReuse) {
+  grpc_ssl_pem_key_cert_pair signed_client_key_cert_pair = {client_key,
+                                                            client_cert};
+  grpc_channel_credentials* client_creds = grpc_ssl_credentials_create(
+      ca_cert, &signed_client_key_cert_pair, nullptr, nullptr);
+
+  do_round_trip(cq, server, server_addr.c_str(), cache, client_creds, false);
+  do_round_trip(cq, server, server_addr.c_str(), cache, client_creds, true);
+  do_round_trip(cq, server, server_addr.c_str(), cache, client_creds, true);
+
+  grpc_ssl_session_cache_destroy(cache);
+
+  GPR_ASSERT(grpc_completion_queue_next(
+                 cq, grpc_timeout_milliseconds_to_deadline(100), nullptr)
+                 .type == GRPC_QUEUE_TIMEOUT);
+
+  grpc_server_shutdown_and_notify(server, cq, grpc_core::CqVerifier::tag(1000));
+  grpc_event ev;
+  do {
+    ev = grpc_completion_queue_next(cq, grpc_timeout_seconds_to_deadline(5),
+                                    nullptr);
+  } while (ev.type != GRPC_OP_COMPLETE ||
+           ev.tag != grpc_core::CqVerifier::tag(1000));
+  grpc_server_destroy(server);
+  grpc_channel_credentials_release(client_creds);
+  grpc_slice_unref(cert_slice);
+  grpc_slice_unref(key_slice);
+  grpc_slice_unref(ca_slice);
+
+  grpc_completion_queue_shutdown(cq);
+  drain_cq(cq);
+  grpc_completion_queue_destroy(cq);
+}
+
+TEST(H2SessionReuseTest, ConcurrentReuse) {
+  int port = grpc_pick_unused_port_or_die();
+
+  std::string server_addr = grpc_core::JoinHostPort("localhost", port);
+
+  grpc_completion_queue* cq = grpc_completion_queue_create_for_next(nullptr);
+  grpc_ssl_session_cache* cache = grpc_ssl_session_cache_create_lru(16);
+
+  grpc_server* server = server_create(cq, server_addr.c_str());
+
+  // create client creds
+  grpc_slice ca_slice, cert_slice, key_slice;
+  GPR_ASSERT(GRPC_LOG_IF_ERROR("load_file",
+                               grpc_load_file(CA_CERT_PATH, 1, &ca_slice)));
+  GPR_ASSERT(GRPC_LOG_IF_ERROR(
+      "load_file", grpc_load_file(CLIENT_CERT_PATH, 1, &cert_slice)));
+  GPR_ASSERT(GRPC_LOG_IF_ERROR("load_file",
+                               grpc_load_file(CLIENT_KEY_PATH, 1, &key_slice)));
+  const char* ca_cert =
+      reinterpret_cast<const char*> GRPC_SLICE_START_PTR(ca_slice);
+  const char* client_cert =
+      reinterpret_cast<const char*> GRPC_SLICE_START_PTR(cert_slice);
+  const char* client_key =
+      reinterpret_cast<const char*> GRPC_SLICE_START_PTR(key_slice);
+  grpc_ssl_pem_key_cert_pair signed_client_key_cert_pair = {client_key,
+                                                            client_cert};
+  grpc_channel_credentials* client_creds = grpc_ssl_credentials_create(
+      ca_cert, &signed_client_key_cert_pair, nullptr, nullptr);
+
+  do_round_trip(cq, server, server_addr.c_str(), cache, client_creds, false);
+  grpc_completion_queue* cq2 = grpc_completion_queue_create_for_next(nullptr);
+
+  std::thread* t1 = new std::thread([&]() -> void {
```

Comment:
nit: Please use full names for variables, so e.g. replace `t1` with `thread_1` and similarly below.